src/partisan/irods.py

```diff
@@ -2438,17 +2650,98 @@ def get(self, local_path: Union[Path, str], **kwargs):
+                objects. See DataObject.put() for more information.
+            local_checksum: A callable that returns a checksum for a local file. See
+                DataObject.put() for more information. This is called for each file in
+                encountered while recursing, with the file path as its argument.
+                (Also accepts a string or a path to a file containing a string, as does
+                DataObject.put(), however this is not useful for collections except in
+                the edge where all the files have identical contents).
+            compare_checksums: Compare caller-supplied local checksums to the remote
+                checksums calculated by the iRODS server after the put operation for
+                data objects. If the checksums do not match, raise an error. See
+                DataObject.put() for more information.
+            fill: Fill in missing data objects in iRODS. If the data object already
+                exists, the operation is skipped. See DataObject.put() for more
+                information.
+            force: Overwrite any data objects already present in iRODS.
             timeout: Operation timeout in seconds.
             tries: Number of times to try the operation.
+
+        Returns:
+            The Collection.
         """
-        raise NotImplementedError()
+        if not Path(local_path).is_dir():
+            raise ValueError(f"Local path '{local_path}' is not a directory")
+
+        self.create(exist_ok=True, timeout=timeout, tries=tries)
+
+        if recurse:
+            for root, dirs, files in os.walk(local_path):
+                for d in dirs:
+                    p = Path(root, d)
+                    r = PurePath(self.path, p.relative_to(local_path))
+                    log.debug(f"Creating collection", local_path=d, remote_path=r)
+                    Collection(r).create(exist_ok=True, timeout=timeout)
+                for f in files:
+                    p = Path(root, f)
+                    r = PurePath(self.path, p.relative_to(local_path))
+                    log.debug(f"Putting data object", local_path=p, remote_path=r)
+                    DataObject(r).put(
+                        p,
+                        calculate_checksum=calculate_checksum,
+                        verify_checksum=verify_checksum,
+                        local_checksum=local_checksum,
+                        compare_checksums=compare_checksums,
+                        fill=fill,
+                        force=force,
+                        timeout=timeout,
+                        tries=tries,
+                    )
+        else:
+            for p in Path(local_path).iterdir():
+                if p.is_dir():
+                    r = PurePath(self.path, p.relative_to(local_path))
+                    log.debug(f"Creating collection", local_path=p, remote_path=r)
+                    Collection(r).create(exist_ok=True, timeout=timeout)
```

Comment:
Does it make sense to archive an empty folder when you're not recursing? Seems a little futile.

Reply:
It makes reasoning about what happens easier; exact mirror versus mirror, but with an exception.